salt/utils/data.py

```diff
@@ -1541,3 +1541,61 @@ def _append_placeholder(value_dict, key):
         else:
             return [{"value": default if obj is not None else obj}]
     return res
+
+
+@jinja_filter("flatten")
+def flatten(data, levels=None, preserve_nulls=False):
```

Comment:
**question** Should this try to detect reference cycles? e.g. with:

```
x = [1,2,3]
x.append(x)
```

Trying to flatten x will currently fail with a RecursionError. Is that correct behavior, or do we want a different behavior?

Reply:
hmmm... good point. This was lifted pretty faithfully from an existing [Ansible filter](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/filter/core.py#L455), so I didn't put a whole lot of thought into the logic.

@waynew , do you have a suggestion for how to guard against that?

Reply:
```
def flatten(data, levels=None, preserve_nulls=False, _ids = None):
    if _ids is None:
        _ids = set()
   if id(data) in set():
         "refence circle!" # raise error or return something like [f"Ref Circle"]

   _ids.add(id(data))
   ...
                 ...  = flatten(data, levels, preserve_nulls, _ids):
   ...
```